### PR TITLE
plugins: Introduce generated, static core plugin registry

### PR DESCRIPTION
**Pull request by sdboyer**
**What this PR does / why we need it**:

This PR introduces a static, codegen'd plugin registry for core plugins. It is a follow-up to #53754, which introduced the `pfs` package that scans and validates an `io/fs` containing plugin meta files (`plugin.json` and `models.cue`). Basically, we get a central static registry of `*pfs.Tree`, one for each core plugin, that can be imported by external repositories, and is kept up to date invisibly and automatically every time anyone runs `make gen-cue`. (And in CI.)

Most surprising for folks will be that this PR introduces Go files under `public/app/plugins`, which historically has been a frontend-only area. There are two basic reasons for doing this:

* It steps up the approach of treating core plugins the same as external plugins begun in #53754.
* It's the simplest possible approach.

The union of these facts means that codegen won't need to change much when we write similar functionality into tools for external plugins. (c.f. this [design doc section](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.elk2cypppbnb)) More details in [this slack thread](https://raintank-corp.slack.com/archives/C01C4K8DETW/p1659430436734929) and linked threads, if any Grafanistas care to dig in.

Once again, there's a lot of changes in here. I'll break it down by path pattern in decreasing order of significance for reviewers:

* `pkg/plugins/pfs/registry/*`: a combination statically populated and handwritten "registry" of core plugins, with particular focus on looking up lineages by slot. As with the `pfs` package in the prior PR, i'm hoping this is useful for the overall plugins system going forward - i.e., `registry.List` can represent `*pfs.Tree`s that come from external plugins just as easily as the core ones.
* `public/app/plugins/*/*/pfs_gen.go`: Embeds the `plugin.json` for that plugin and makes it available in Go via a func named `PluginTree()`. References to each plugin's generated `PluginTree()` func are what's generated into `pkg/plugins/pfs/registry`.
  * If the plugin contains a `models.cue`, that's also embedded, and standard Thema LineageFactory functions are generated to get at the lineage. (`*pfs.Tree` serves our needs though so we don't use them; i may ultimately choose to not to generate them at all.)
* `pkg/codegen/pluggen.go`: the meat of the new codegen logic, which produces everything else of substance in this PR.
* `pkg/codegen/**` (others): codegen framework refactoring. Really needed to do this to make all this feasible. I could try to separate it out, but it would be a PITA - it would be amazing if kind reviewers could just ignore it.

Everything else not in this list is noise related to the codegen refactor. It improves, or at least makes consistent, a lot of comments - but ultimately nothing that impacts behavior. (`pkg/codegen` is rapidly moving towards a framework. On balance, that's a good thing, even if it means noise right now!)

Merging this PR completes most of [pass 4 of the schematization design](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.sc1cwrlnanla). What remains is composition in Thema itself (basically item 2 on my TODO list now), and implementing the 'dist' registry. I didn't bother with the latter because it's not strictly necessary for me to do codegen in grok. Which is what's next - taking the generated registry here, along with the existing coremodel registry, and getting grok properly ready for public use.

**Special notes for your reviewer**:

Note that everything here also works for datasource plugins, though none of the datasource plugins are converted yet.

Currently, we do not generate Go types for panel plugins. That's because i haven't yet dealt with converting [imports of our special CUE package](https://github.com/grafana/grafana/blob/4dbe0b4f025f65e07b168f306def5683356fb5d1/public/app/plugins/panel/barchart/models.cue#L19) to a Go equivalent.

We _do_ generate types for datasource plugins, but since no datasource plugins have implemented schema yet, i've made [this gist](https://gist.github.com/sdboyer/4f79e228fd5c1e9924b970c8ba60510b) to illustrate. That's also worth having a look at because it shows how this whole system behaves in the presence of multiple slot implementations/lineages, which we have for datasources. (Yes, there's a bug with generating Go code for `DSOptions`, but IMO it can be fixed in a follow-up)

Note that there really isn't an equivalent to the generated `coremodel.Interface` (e.g. [here](https://github.com/grafana/grafana/blob/main/pkg/coremodel/pluginmeta/pluginmeta_gen.go#L564-L584)) for plugins. The closest thing is `*pfs.Tree` and just plain old `"github.com/grafana/thema".Lineage` itself. I think that's good enough for us to start with.

### CODE REVIEW

**Comment by sdboyer** [comment]
cc @grafana/user-essentials @grafana/observability-experience-squad @grafana/observability-metrics @grafana/observability-logs @grafana/aws-plugins @grafana/enterprise-datasources @grafana/cloud-provider-plugins @grafana/grafana-bi-squad all of you folks maintain plugins. This PR is going to make this `pfs_gen.go` file show up in a place that you will, no doubt, find surprising. I figured y'all deserve a heads-up, at least :) See OP for explanation.

cc @ying-jeanne @sakjur @yuri-tceretian - interesting examples here of what extending the codegen system looks like. And also, how it's becoming more "system-ish" in general

cc @IfSentient reiterating what i said in slack, the codegen framework patterns that are starting to emerge here are ones we'll likely want to push down into Thema, or at least reusing in some of your sdk contexts

**Comment by ryantxu** [comment]
I am still trying to understand the need/reason/motivation for this.  It may just be a few steps ahead of where we are.  Is the goal to revisit how we handle the static plugin registry?  @leventebalogh has some docs discussing possible directions.

What are we solving?  or is this just proving that thema could be used for more things?

File: public/app/plugins/panel/barchart/pfs_gen.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+var parseOnce sync.Once
+var ptree *pfs.Tree
+
+//go:embed plugin.json models.cue
+var plugFS embed.FS
+
+// PluginTree returns the plugin tree representing the statically analyzable contents of the barchart plugin.
+func PluginTree(lib *thema.Library) *pfs.Tree {
```
Comment (ryantxu) [inline public/app/plugins/panel/barchart/pfs_gen.go:40]: what are these files, and whey do we need them?

**Comment by sdboyer** [comment]
It seems to me that this is all answered in the PR description, as well as its predecessor PR. What is unclear? 

the existing static plugin registry is not externally importable, not complete (has no panel plugin types, because those don't exist at all), offers no access to plugin declared lineages.

that makes it unsuitable for either of the two critical path goals:

* actually compose plugin schemas into coremodel schemas
* generate types for grok, the foundation of all as-code experiences we want to build there

and crucially, do it all automatically, so that manual maintenance is not a part of the process. The clear evidence, for years, is that it does not get done when it's manual, and as-code remains a second class citizen. I had two conversations just this week reinforcing this.

This PR is the final essential piece of the conveyer belt for the above goals. And, happily, several others.

File: public/app/plugins/panel/barchart/pfs_gen.go (hunk 1 above)
Comment (sdboyer) [reply public/app/plugins/panel/barchart/pfs_gen.go:40]: Please read the PR description.

**Comment by sdboyer** [comment]
Also, for clarity, this PR (and its two predecessors, #53754 and #52121) is the expected outgrowth of weeks of conversation between @wbrowne and i. I don't think this is a surprise to plugins folks :)

**Comment by sdboyer** [comment]
given how much the stuff in `pkg/plugins/pfs/registry` kinda isn't a registry, but is really just a list...i'm thinking about renaming the package to `list` or something.

Or maybe putting the base `List` type directly in `pfs`, then renaming `registry` to `corelist`.

**Comment by sdboyer** [comment]
`registry` definitely needs to be a separate package from `pfs` - you should be able to import the parser without having all the embedded core plugins come along for the ride - but yeah, the current arrangement isn't great.

🤔

**Comment by sdboyer** [comment]
OK! talked with @ryantxu a bit, and he showed me how the `pfs_gen.go` pattern diverges pretty sharply from the existing way that external plugins usually get built.

That's not _necessarily_ a problem, and it's plausible we end up with something like that pattern. Things will have to change in the quest for core/external plugin uniformity, after all. But i also thought of:

* Some reasonable ways of allowing plugins uniform control over how the gen works, in the future
* A reasonable path to achieving the same statically generated effect without needing to actually import the `pfs_gen.go` files, and therefore not need to generate them at all

So, i stopped generating `pfs_gen.go` completely for all plugins, and we can hack away at more standardized codegen before reinstating it. 

Panel plugins will definitely be fine without those generated lineage/pfs bindings. Core datasources might want them, but we have enough time before datasource plugin maintainers start implementing them to make progress on the standardized codegen.

So, this PR is now 2500 lines lighter :)

File: .gitignore [hunk 2]
```diff
@@ -165,6 +165,8 @@ compilation-stats.json
 !pkg/services/featuremgmt/toggles_gen.go
 !pkg/coremodel/**/*_gen.go
 !pkg/framework/**/*_gen.go
+!pkg/plugins/pfs/**/*_gen.go
```
Comment (ryantxu) [inline .gitignore:168]: still needed?

Comment (sdboyer) [reply .gitignore:168]: this one is, there's still the generated "registry"

**Comment by sdboyer** [comment]
@yuri-tceretian given the consistent pathing issues with windows, would you mind giving this one a poke?

we, or at least i, clearly really need a windows test environment 😢

**Review by wbrowne — APPROVED** [review]

**Pull request merged**